### ISSUE TITLE
Optimize alignment process

### ISSUE THREAD

**Issue by mezzodrinker** [mezzodrinker/spas#11]
Currently, when updating the rotation direction, the script runs the current rotor and checks whether the current rotation direction yields a higher output than before. If that is not the case, the script reverses the rotor and carries on. To optimize the alignment process, though, the script should check *both* directions for their respective increase in power, compare those, and then rotate into the direction whose change in power was closest to positive infinity. For example:

> The script checks into which direction it should rotate. The old direction yields a power change of -1kW, the opposite direction yields a power change of -5kW. Because -1kW is closer to positive infinity than -5kW, the script keeps the current rotation direction.

Also, to prevent the script from permanently checking while still aligning the panels, the script should not check for directions again until the change in power went above 0.

**Comment by kennumen** [comment]
In a single rotor system, is this not an appropriate place to idle? If 'left' is worse and 'right' is worse, haven't we reached optimal position?

Along those same lines, in a multiple rotor system, that may be a good time to alternate rotors (status partial-temp-idle or precision-aligning) and see which, if any, can provide better results with further alignment. Obviously if one does get better then the other rotor/axis probably needs further alignment as well.

**Comment by mezzodrinker** [comment]
> In a single rotor system, is this not an appropriate place to idle? If 'left' is worse and 'right' is worse, haven't we reached optimal position?

You're right, I was completely focused on the multi-rotor setup there and didn't even think about just using one rotor, thus never tested it. I should add that exception to my local code, though I doubt there will be much of a difference...

> Along those same lines, in a multiple rotor system, that may be a good time to alternate rotors (status partial-temp-idle or precision-aligning) and see which, if any, can provide better results with further alignment. Obviously if one does get better then the other rotor/axis probably needs further alignment as well.

So you mean instead of checking both directions of _one_ rotor I should check both directions of _all_ rotors? Sounds like this may fix the problem, but with the (sadly) quite limited resources I get from Keen this might turn out to be a coding hell...

**Comment by kennumen** [comment]
Okay I had a whole wall of text written out but decided better of it and tried to keep it short.

With the second part, I'm focusing on two rotor systems. 3 or more rotors for alignment functionally add little while adding quite a lot of complexity.

To quote the above example, where one side is -1 kW and the other is -5 kW. This means one of two things: either you're reversing a single rotor which goes from -2 kW change to +2 kW change and the other one stays at -3 kW change (this rotor is off or not reversed), OR you're reversing both rotors where one rotor rotor again has either -2 kW change or +2 kW change and the other should be idling because either side creates a -3 kW change.

You could use your current panel naming trick on the rotors (giving each its own status?) and even with the current system check what change the 'active/testing/aligning/...' is doing  - one rotor at a time in a round-robin manner. After all a change of 0 kW could mean one's changing +2 but the other's changing -2.

If you can't check both sides in the same program loop, one of the data will be outdated though possibly not significantly.

In short:
- You need to know per rotor how moving it is affecting it - ideally but not necessarily a value for each side. If combined with individual rotor idling, you <probably> don't even need to stop the other rotor(s) assuming loops happen often enough (the more rotors, the more loops <during alignment/testing; excluding rotors hibernating=idling>?).
- You need to be able to idle any rotor individually once changes are 'insignificant'. This might replace a 'target/idling output' (e.g. 129.5f) in systems with a single rotor. You would need to know **exactly** what this rotor is doing to the change so this would mean stopping the other rotors. I've been playing with small ship solar panels but insignificant is one **positive** change below roughly 1%? of the max panel output (meaning 30 kW or below 1% being below 0.30 kW, for these small ship panels).
- Once all rotors are idling (and system is not in hibernate status) you only need to align one axis at a time, since you're just tracking the sun now and alignment changes should be minimal.

All that should also have the added benefit of making the whole system look a lot calmer as well, even on setups where idling output can't be reached.

**Comment by mezzodrinker** [comment]
To summarize your summary, you want me to treat each rotor in each axis individually and - once all rotors in an axis are not contributing any significant changes in power anymore - go to the next axis?

**Comment by kennumen** [comment]
Before I say anything else, could you clarify if variables can be passed to the next loop or if you start each loop 'fresh'? Because I'm still assuming "no" since you store oldPower in the solar panel name and this is a fairly crippling limitation.

Well there's ideal theory and then there's what you can do with what Keen will let you do. Fact is if two or more rotors are moving you don't know what each rotor's movement is doing to the change, only what their combined movement is doing. It's also a fact that this is a relatively complicated process compared to simply moving both axis in tandem. 

I'm completely ignoring 3+ axis here since the more rotors the tougher it is to get a single rotors effect from the general change. 

I'm bad at putting things into words so I often use too many. Also I'm not saying this is simple or effortless, but here's what would probably work best in simple terms:
- while changes are "big" (greater-than value TBD) you can assume both rotors have same effect and keep both same or reverse both.
- if changes are "small" (range TBD) you need specifics.
  - or they are counter acting each other
  - or they are both just small
    - one or both may be near (individual rotor) idling limit
  - or both?

If changes are small, more fleshed out:
- turn one rotor off (just one loop), look at change of individual rotor in relation to general data, deduce **approximate** change of the other one (comparing current loop individual to previous loop general values => muddy data) and turn the other rotor back on.
  - if one is or both are near (individual rotor) idling limit (less-than value TBD, 1% of design max output?) -> ROTOR_STATUS = IDLE
    - with testing decide whether to add a "half rotation speed" period (probably not)
  - if one is or both are negative -> ROTOR_REVERSE _note: could it be necessary to add a negative idling limit?_
  - else it must be positive yet still 'significantly' improving -> do nothing
- with testing decide whether to let all rotors (also being the "general" value) on for just the 1 loop or 2/3/4... Perhaps an extra short (panic mode?) loop when only one rotor is running. Plenty of simple (potentially) optimization ideas.

Now three or more axis are more complex. You can't assume the last value (since it the data after n-1 loops would be invalid). Beyond that the process is very similar. If changes are "big" and negative, reverse all. Else turn off all but one rotor (looping through the rotor list, ignoring hibernating/idling ones), evaluate just this rotor. Idle it if necessary (longer idle the more rotors there are?), reverse if applicable else leave it be.

At 4 or more axis it might just be preferable to let it spasm its way to perfection...
